jbmc/src/java_bytecode/README.md

```diff
@@ -41,9 +41,32 @@ To be documented.
 
 To be documented.
 
-\subsection java-bytecode-genereic-substitution Generic substitution
-
-To be documented.
+\subsection java-bytecode-generic-substitution Generic substitution
+
+Generic types are resolved at runtime in Java. Therefore, when we initialize
+variables that have generic type we may want to dynamically replace the generic
+parameters with their specialized types rather than using only the raw type. For
+example, consider a generic class `MyGeneric<T>` with a field `T myField`. When
+initializing a variable `MyGeneric<Integer> mg` we change the type of its field
+`mg.myField` from `MyGeneric::T` to `Integer`. Generic substitution is used
+in pointer type selection (see \ref java-bytecode-pointer-type-selection).
+
+The generic substitution (or specialization) relies on a map that stores the
+concrete types of all generic parameters in the current scope. The map is
+maintained in \ref java_object_factoryt::generic_parameter_specialization_map.
+Note that not all generic parameters need to have a concrete type, e.g., when
+the method under test is generic. The types are removed from the map when the
+scope changes. In different depths of the scope the parameters can be
+specialized with different types so we keep a stack of types for each parameter.
+
+We use the map in \ref select_pointer_typet::specialize_generics to
+retrieve the concrete type of generic parameters such as `MyGeneric::T` and of
+arrays of generic parameters such as `MyGeneric::T[]`. A special attention
+is paid to breaking infinite recursion when searching the map, e.g.,
+`MyGeneric::T` being specialized with `MyOtherGeneric::U` and vice versa. More
+complicated generic types such as `MyGeneric<T>` are specialized indirectly
```

Comment:
It might be worth explaining the thing about the more complicated types in the documentation of `select_pointer_typet::specialize_generics`? I looked at that function for the first time while reviewing your other PR and was wondering if those more complicated types were a missing case or if they had been left out for a reason.

Reply:
I added the explanation to the function doc in the other PR https://github.com/diffblue/cbmc/pull/3382/commits/75cb991d82276fb32bda00c6b55c3b8d06436731